server/auth/auth.go

```diff
@@ -8,11 +8,61 @@ import (
 	"io/ioutil"
 	"net/http"
 	"strings"
+	"sync"
 	"time"
 
+	"github.com/gorilla/context"
 	"github.com/pinterest/knox"
 )
 
+type contextKey int
+
+const (
+	principalCtxKey contextKey = iota
+)
+
+type PrincipalContext interface {
+	SetCurrentPrincipal(principal knox.Principal)
+	GetCurrentPrincipal() knox.Principal
+}
+
+type principalContext struct {
+	principalCtxKey contextKey
+	request         *http.Request
+	once            sync.Once
+	invocationCount int
+}
+
+func NewPrincipalContext(request *http.Request) PrincipalContext {
+	return &principalContext{
+		principalCtxKey,
+		request,
+		sync.Once{},
+		0,
+	}
+}
+
+func (ctx *principalContext) GetCurrentPrincipal() knox.Principal {
+	if rv := context.Get(ctx.request, principalCtxKey); rv != nil {
+		return rv.(knox.Principal)
+	}
+	return nil
+}
+
+func (ctx *principalContext) SetCurrentPrincipal(principal knox.Principal) {
+	if ctx.invocationCount > 0 {
+		panic("SetPrincipal was called more than once during the lifetime of the HTTP request")
+	}
```

Comment:
do we really want the server to crash if this happens?

Reply:
I thought it was better to explicitly fail because someone might inadvertently end up calling `SetPrincipal` twice, wonder why their changes aren't taking effect and then get confused. By having an explicit panic, with an explicit error message, one would understand that they are in fact, doing something wrong

The other thing we could do is allow `SetPrincipal` to return an error (vs just panic) but there's no easy way to log that AFAICT.